content/match-types-spec.md

```diff
@@ -0,0 +1,354 @@
+Unfortunately, the presence of the abstract type constructor case makes that impossible.
+Indeed, because of it, a pattern that is legal at definition site may become illegal after some later substitution.
+
+Consider for example the standard library's very own `Tuple.InverseMap`:
+
+```scala
+/** Converts a tuple `(F[T1], ..., F[Tn])` to `(T1,  ... Tn)` */
+type InverseMap[X <: Tuple, F[_]] <: Tuple = X match {
+  case F[x] *: t => x *: InverseMap[t, F]
+  case EmptyTuple => EmptyTuple
+}
+```
+
+If we instantiate `InverseMap` with a class type parameter, such as `InverseMap[X, List]`, the first case gets instantiated to
+```scala
+case List[x] *: t => x *: InverseMap[t, List]
+```
+which is legal.
+
+However, nothing prevents us a priori to instantiate `InverseMap` with an illegal type constructor, for example
+```scala
+type IsSeq[t <: Seq[Any]] = t
+InverseMap[X, IsSeq]
+```
+which gives
+```scala
+case IsSeq[x] *: t => x *: InverseMap[t, IsSeq]
+```
+
+These instantiatiations happen deep inside the type checker, during type computations.
+Since types are cached, shared and reused in several parts of the program, by construction, we do not have any source code position information at that point.
+That means that we cannot report *warnings*.
+
+We can in fact report *errors* by reducing to a so-called `ErrorType`, which is aggressively propagated.
+This is what we do in the proposed implementation (unless using `-source:3.3`).
+
+### Open questions
+
+None at this point.
+
+## Alternatives
+
+The specification is more complicated than we initially wanted.
+At the beginning, we were hoping that we could restrict match cases to class type constructors only.
+The quantitative study however revealed that we had to introduce support for abstract type constructors and for type member extractors.
+
+As already mentioned, the standard library itself contains an occurrence of an abstract type constructor in a pattern.
+Making that an error would mean declaring the standard library itself bankrupt, which was not a viable option.
+
+We tried to restrict abstract type constructor to never match on their own.
+Instead, we wanted them to stay *stuck* until they could be instantiated to a concrete type constructor.
+However, that led some existing tests to fail even for match types that were declared legal, because they did not reduce anymore in some places where they reduced before.
+
+Type member extractors are our biggest pain point.
+Their specification is complicated, and the implementation as well.
+Our quantitative study showed that they were however "often" used (10 occurrences spread over 4 libraries).
+In each case, they seem to be a way to express what Scala 2 type projections (`A#T`) could express.
+While not quite as powerful as type projections (which were shown to be unsound), match types with type member extractors delay things enough for actual use cases to be meaningful.
```

Comment:
@dwjinand you might have something to say about this? I know it's come up repeatedly in our discussions about match types that it's possible to use them to smuggle in something akin to abstract type projections.

Reply:
> something akin to abstract type projections

With the proposed spec, only *concrete* type projections actually make it. As long as it's abstract, `ExtractTOf[A]` cannot reduce to `T` like `A#T` does. The difference is that `A#T` has the "best" *bounds* of `T` in `A`, whereas `ExtractTOf[A]` has only the match type's bound as long as `T` is not concrete (through subsequent substitution of `A` by something more precise in which `T` is a type alias whose rhs does not depend on its prefix).

For example,
```scala
class Base {
  type T
  def someT: T = ???
}

type ExtractTOfAux[A] = Base { type T = A }
type ExtractTOf[X <: Base] = X match
  case ExtractTOfAux[t] => t

// Scala 2 abstract type projection, works
def fooOfBaseProj[B <: Base](b: B): B#T = b.someT

// Scala 3 match type extractor: does not work because the match type does not reduce
// as T is not concrete in B.
def fooOfBaseMT[B <: Base](b: B): ExtractTOf[B] = b.someT
```

The match-type-based type member extractor still features enough power because you can "carry around" values of that type, and at the edge your `B` becomes concrete enough that you can reduce. This is what happens with `scala.Enumeration`s.

IMO, people should not use that. It's all a hack. Instead, we should use path-dependent types with actual values of the enclosing class. For example the proper definition of `fooOfBase` is:
```scala
def fooOfBasePathDep(b: Base): b.T = b.someT
```
Unfortunately, people have been using type projections/type member extractors instead. Yes, sometimes it avoids one extra layer of nesting: if you need a class that `extends Foo[b.type]`, you would `b` to be declared one level up, and so that class needs to be nested in another class that has a `b` field. This is why the json4s workaround is what it is, instead of turning it into path-dependent types. The correct solution would be to have
```scala
class EnumSerializers(val enumeration: Enumeration) {
  class EnumSerializer extends Serializer(enumeration.Value) {
    ...
  }
}
```
but that would have changed the public API, so it was not an acceptable PR to make.